### PR TITLE
Button | HTML | Add missing 'disabled' attribute to button and trigger

### PR DESCRIPTION
**Pull request by danielamorse**
## Jira

http://vjira2:8080/browse/BDS-1563

## Summary
Add `disabled` attribute to `bolt-button` and `bolt-trigger` inner elements when prop is set.

## Details

Bolt button and trigger components do not currently add the `disabled` attribute to the inner semantic `<button>` element even though they receive the proper visual style. Without the `disabled` attribute `<button>` is still focusable according to the Accessibility Object Model.

## How to test

- Given I am running the `fix/add-button-disabled-attribute` branch locally
- When I go to the [Button style variation page](http://localhost:3000/pattern-lab/patterns/02-components-button-15-button-style-variations/02-components-button-15-button-style-variations.html) and I inspect each "disabled" button example
- Then I see the `disabled` attribute on each component's inner `<button>` tag
- When I go to the [Trigger disabled variation page](http://localhost:3000/pattern-lab/patterns/02-components-trigger-30-trigger-disabled-variations/02-components-trigger-30-trigger-disabled-variations.html) and I inspect each "disabled" trigger example
- Then I see the `disabled` attribute on the first example's inner element and `aria-disabled="true"`  on the second, and the second does not have an `href` attribute
- When I paste the following JS into the console and I inspect the four web components added to the page:
```
var el = document.createElement('div');
el.innerHTML = `<div><bolt-button disabled>Basic button (disabled)</bolt-button></div>
                <div><bolt-button disabled url="http://pega.com">Basic button link (disabled)</bolt-button></div>
                <div><bolt-trigger disabled>Basic trigger (disabled)</bolt-trigger></div>
                <div><bolt-trigger url="http://pega.com" disabled>Basic trigger link (unchanged by 'disabled' prop)</bolt-trigger></div>`;
document.body.appendChild(el);
```
- Then I see the `disabled` attribute on the inner `<button>` tags and `aria-disabled="true"` on the inner `<a>` tags, and `<a>` tags do not have `href` attributes

### CODE REVIEW

**Comment by danielamorse** [comment]
> I’d recommend we add `aria-disabled=“true”` to rendered links and _possibly_ remove the href (temporarily) if a link is `disabled`.

@sghoweri @mikemai2awesome, I've applied the disabled state to `bolt-button` and `bolt-trigger` when they render *links*. Instead of getting the `disabled` prop like buttons do, they get `aria-disabled="true"` and the `href` is removed. The `href` is re-added if the `disabled` prop is removed. I've updated the examples and testing steps. Ready for re-review.

**Comment by danielamorse** [comment]
> Just double check one thing, is this just a browser glitch? `disabled="true"` is rendered instead of just `disabled` in the `<button>` element.
> 

@mikemai2awesome This is not a browser glitch. Twig renders `disabled` and our JS renders `disabled="true"`. For consistency I updated the JS to render `disabled` ([see commit](https://github.com/bolt-design-system/bolt/pull/1260/commits/aa28e34b470659eaa927c391c4ec910e841850cc)). Good call.

**Comment by mikemai2awesome** [comment]
@danielamorse Awesome, glad I pointed it out.

File: packages/components/bolt-button/__tests__/button.js [hunk 1]
```diff
@@ -62,6 +62,25 @@ describe('button', () => {
...
+    expect(results.html).toMatchSnapshot();
+  });
+
+  test('Button with "disabled" adds attr to <a>', async () => {
+    const results = await render('@bolt-components-button/button.twig', {
+      text: 'This is a button',
+      url: 'http://pega.com',
+      disabled: true,
+    });
+    expect(results.ok).toBe(true);
```
Comment (sghoweri) [inline packages/components/bolt-button/__tests__/button.js]: @danielamorse food for thought for your next PR similar to this one: 

While snapshots are good quick way to at least yell at us when HTML changes, they aren’t a foolproof way to confirm this prop continues to work as expected.

Ideally we’d _also_ be peeking into the rendered HTML (Twig and/or WC) to actually check that the HTML element itself has the attribute or class we’re expecting, similar to what I’m doing for the Table component: https://github.com/bolt-design-system/bolt/blob/1d61853d036a69c364c7a458d03f96d4071b316c/packages/components/bolt-table/__tests__/table.js#L173

File: packages/components/bolt-button/src/button.js [hunk 2]
```diff
@@ -145,23 +145,43 @@ class BoltButton extends BoltAction {
...
+        if (this.props.disabled) {
+          buttonElement.setAttribute('disabled', '');
+        } else {
+          buttonElement.removeAttribute('disabled');
+        }
       }
 
       if (this.props.tabindex) {
-        buttonElement.setAttribute('target', this.props.tabindex);
+        buttonElement.setAttribute('tabindex', this.props.tabindex);
```
Comment (sghoweri) [inline packages/components/bolt-button/src/button.js]: +1 good catch

File: packages/components/bolt-button/src/button.js [hunk 3]
```diff
@@ -145,23 +145,43 @@ class BoltButton extends BoltAction {
       buttonElement = this.rootElement.firstChild.cloneNode(true);
       buttonElement.className += ' ' + classes;
 
-      if (this.props.url) {
-        buttonElement.setAttribute('href', this.props.url);
-      }
-
-      if (this.props.target) {
-        buttonElement.setAttribute('target', this.props.target);
+      if (buttonElement.tagName === 'A') {
```
Comment (sghoweri) [inline packages/components/bolt-button/src/button.js]: I’m good with this entire block of changes: I just wish / think we could really use a way to automatically dissolve original HTML elements into their respective component props + custom attributes added on.

**Comment by danielamorse** [comment]
> Just one big questions I want to make sure we’ve double-checked before OKing this (and having you merge this down to master and update release notes):
> 
> 1. If a button / trigger had it’s href set via the attributes object (not url prop), this doesn’t get passed along to the originally rendered link, right?
> 2. If ^ did still get passed along to the rendered HTML tag (but not added as a prop) and was set to disabled... would this URL even show up at all in the rendered Twig template?
> 
> LMK if this makes sense!!

@sghoweri, makes sense. `button` and `target` Twig templates do nothing to handle `href`, whether it comes from props or attributes. Only `link` Twig template has logic to handle `href` (though it prefers `url`). So, you are right. Nothing gets passed along if you use `href`. I'm ok with that since the API says `url`, and I don't think we want the overhead of supporting an invalid prop, except maybe in `link` component.

I went ahead and added "todos" for each of your test and template suggestions. Thanks for the feedback!

File: packages/components/bolt-button/__tests__/button.js (hunk 1 above)
Comment (danielamorse) [reply packages/components/bolt-button/__tests__/button.js]: @sghoweri added todo 👍

**Pull request merged**